server/coordinator/procedure/manager_impl.go

```diff
@@ -0,0 +1,191 @@
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *ManagerImpl) SubmitMigrateTask(_ context.Context, _ *MigrateRequest) error {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *ManagerImpl) SubmitSplitTask(_ context.Context, _ *SplitRequest) error {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *ManagerImpl) SubmitMergeTask(_ context.Context, _ *MergeRequest) error {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *ManagerImpl) Cancel(ctx context.Context, procedureID uint64) error {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	for _, procedure := range m.procedures {
+		if procedure.ID() == procedureID {
+			err := procedure.Cancel(ctx)
+			return errors.WithMessagef(err, "cancel procedure failed, procedureID = %d", procedureID)
+		}
+	}
+	return nil
+}
+
+func (m *ManagerImpl) ListRunningProcedure(_ context.Context) ([]*RunningProcedureInfo, error) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	procedureInfos := make([]*RunningProcedureInfo, 0)
+	for _, procedure := range m.procedures {
+		procedureInfos = append(procedureInfos, &RunningProcedureInfo{
+			ID:    procedure.ID(),
+			Typ:   procedure.Typ(),
+			State: procedure.State(),
+		})
+	}
+	return procedureInfos, nil
+}
+
+type RunningProcedureInfo struct {
+	ID    uint64
+	Typ   Typ
+	State State
+}
+
+func NewManagerImpl(ctx context.Context, cluster *cluster.Cluster, client *clientv3.Client, rootPath string) (Manager, error) {
+	manager := &ManagerImpl{
+		storage:              NewEtcdStorageImpl(client, cluster.GetClusterID(), rootPath),
+		cluster:              cluster,
+		dispatch:             eventdispatch.NewDispatchImpl(),
+		shardIDAllocator:     id.NewReusableAllocatorImpl(make([]uint64, 0), 0),
+		procedureIDAllocator: id.NewAllocatorImpl(client, "procedure", 5),
+	}
+
+	err := manager.Init(ctx)
+	if err != nil {
+		return nil, errors.WithMessage(err, "init manager failed")
+	}
+
+	return manager, nil
+}
+
+func (m *ManagerImpl) Init(ctx context.Context) error {
+	procedureChan := make(chan Procedure, 10)
+	m.procedureChan = procedureChan
+	go startProcedureWorker(ctx, procedureChan, make(chan error, 10))
+	if err := m.SubmitScatterTask(ctx, &ScatterRequest{}); err != nil {
+		return errors.WithMessage(err, "submit scatter task failed")
+	}
+	err := m.retryAll(ctx)
+	if err != nil {
+		return errors.WithMessage(err, "retry all running procedure failed")
+	}
+	return nil
+}
+
+func (m *ManagerImpl) retryAll(ctx context.Context) error {
+	metas, err := m.storage.List(ctx, 100)
+	if err != nil {
+		return errors.WithMessage(err, "storage list meta failed")
+	}
+	for _, meta := range metas {
+		if !checkNeedRetry(meta) {
+			continue
+		}
+		p := load(meta)
+		err := m.retry(ctx, p)
+		return errors.WithMessagef(err, "retry procedure failed, procedureID = %d", p.ID())
+	}
+	return nil
+}
+
+func startProcedureWorker(ctx context.Context, procedures <-chan Procedure, results chan<- error) {
```

Comment:
Is it better to define this method on `ManagerImpl`? BTW, it seems there are no considerations taken into how to exit such a loop.